topic-operator/src/main/java/io/strimzi/operator/topic/Session.java

```diff
@@ -133,72 +136,102 @@ public void init(Vertx vertx, Context context) {
-                }
-                return Future.succeededFuture();
+                LOGGER.info("Stopped");
+                return Future.future(handle -> stop.complete());
             });
+    }
 
-            promise.future().compose(ignored -> {
+    private Future<Void> closeKubernetesWatch() {
+        LOGGER.debug("Closing Kubernetes watch");
+        if (topicWatch != null) {
+            topicWatch.close();
+        }
+        return Future.succeededFuture();
+    }
 
-                LOGGER.debug("Disconnecting from zookeeper {}", zk);
-                zk.disconnect(zkResult -> {
-                    if (zkResult.failed()) {
-                        LOGGER.warn("Error disconnecting from zookeeper: {}", String.valueOf(zkResult.cause()));
-                    }
-                    long timeoutMs = Math.max(1, deadline - System.currentTimeMillis());
-                    LOGGER.debug("Closing AdminClient {} with timeout {}ms", adminClient, timeoutMs);
-                    try {
-                        adminClient.close(Duration.ofMillis(timeoutMs));
-                        HttpServer healthServer = this.healthServer;
-                        if (healthServer != null) {
-                            healthServer.close();
-                        }
-                    } catch (TimeoutException e) {
-                        LOGGER.warn("Timeout while closing AdminClient with timeout {}ms", timeoutMs, e);
-                    } finally {
-                        LOGGER.info("Stopped");
-                        blockingResult.complete();
+    private Future<Void> stopZooKeeperWatcher() {
+        LOGGER.debug("Stopping ZooKeeper watcher");
+        if (topicsWatcher != null) {
+            topicsWatcher.stop();
+        }
+        return Future.succeededFuture();
+    }
+
+    private Future<Void> stopTopicStoreService() {
+        LOGGER.debug("Stopping topic store service");
+        if (service != null) {
+            service.stop();
+        }
+        return Future.succeededFuture();
+    }
+
+    @SuppressFBWarnings("RV_RETURN_VALUE_IGNORED_BAD_PRACTICE")
+    private Future<Void> disconnectFromZk(ExecutorService shutdownExecutor) {
+        LOGGER.debug("Disconnecting from ZooKeeper");
+        Promise<Void> promise = Promise.promise();
+        if (zk != null) {
+            shutdownExecutor.submit(() -> {
+                zk.disconnect(result -> {
+                    if (result.failed()) {
+                        LOGGER.warn("Error disconnecting from ZooKeeper: {}", String.valueOf(result.cause()));
+                        promise.fail(result.cause());
                     }
+                    promise.complete();
                 });
-                return Future.succeededFuture();
             });
-        }, stop);
+        }
+        return promise.future();
+    }
+
+    @SuppressFBWarnings("RV_RETURN_VALUE_IGNORED_BAD_PRACTICE")
+    private Future<Void> closeKafkaAdminClient(ExecutorService shutdownExecutor) {
```

Comment:
This is very deceptive. This method signature suggests that the close is asynchronous (which it is), but by returning a `Future` we're also implying that that future will be _completed when the `close()` is finished_, and that's *not* the case.

This should either return `void`, or the returned `Future` should complete (on a defined thread) once the close has completed.